### PR TITLE
Integrate `pysa-eur-sec` into the pipeline

### LINKED ISSUE #306: Migrate pypsa-eur-sec to datasets

### PR DESCRIPTION
**Pull request by ulfmueller**
Fixes issue #138.
Closes #306

### CODE REVIEW

**Comment by ulfmueller** [comment]
@gnn the virtualenv Operator seems to work well. But I have troubles with the pathes in the virtualenv. Can you check that out? In particular it runs into an error because it cannot find the path [here](https://github.com/openego/eGon-data/blob/features/%23138-integrate-pysa-eur-sec/src/egon/data/importing/pypsaeursec/__init__.py#L65)

**Comment by ulfmueller** [comment]
These path errors are weird. I don't get it...

```
[2021-04-20 15:21:26,448] {python_operator.py:323} INFO - Got error output
b'Traceback (most recent call last):\n  File "/tmp/venvd8u4gtga/script.py", line 79, in <module>\n    res = run_pypsa_eur_sec(*args, **kwargs)\n  File "/tmp/venvd8u4gtga/script.py", line 52, in run_pypsa_eur_sec\n    subproc.run(\n  File "/tmp/venvd8u4gtga/lib/python3.8/site-packages/egon/data/subprocess.py", line 59, in run\n    result = subprocess.run(*args, **kwargs)\n  File "/usr/lib/python3.8/subprocess.py", line 489, in run\n    with Popen(*popenargs, **kwargs) as process:\n  File "/usr/lib/python3.8/subprocess.py", line 854, in __init__\n    self._execute_child(args, executable, preexec_fn, close_fds,\n  File "/usr/lib/python3.8/subprocess.py", line 1702, in _execute_child\n    raise child_exception_type(errno_num, err_msg, err_filename)\nFileNotFoundError: [Errno 2] No such file or directory: PosixPath(\'pysa-eur-sec/data\')\n'
[2021-04-20 15:21:27,397] {taskinstance.py:1150} ERROR - Command '['/tmp/venvd8u4gtga/bin/python', '/tmp/venvd8u4gtga/script.py', '/tmp/venvd8u4gtga/script.in', '/tmp/venvd8u4gtga/script.out', '/tmp/venvd8u4gtga/string_args.txt']' returned non-zero exit status 1.
```

The repository is there but anyhow it is not found. Any ideas?

**Comment by ulfmueller** [comment]
That path problem was fixed by 75c9e16424e88de9f2c41c0f3b6c276d554aad41 .  A simple typo. Thx @ClaraBuettner for pointing me towards it!

**Comment by ulfmueller** [comment]
@ClaraBuettner could you test if the task "run_pypsaeursec" runs flawless on your machine (considering that you have snakemake, conda and the egon-data stuff installed) some time next week? Probably you have to have installed libtbb2 ( `sudo apt install libtbb2 `)

**Comment by ClaraBuettner** [comment]
I tested this branch but the task failed with the following log: 

> Select jobs to execute...
> |   [Tue May 11 10:25:23 2021]
> |   rule build_shapes:
> |       input: data/bundle/naturalearth/ne_10m_admin_0_countries.shp,
> |   data/bundle/eez/World_EEZ_v8_2014.shp,
> |   data/bundle/NUTS_2013_60M_SH/data/NUTS_RG_60M_2013.shp,
> |   data/bundle/nama_10r_3popgdp.tsv.gz,
> |   data/bundle/nama_10r_3gdp.tsv.gz, data/bundle/ch_cantons.csv,
> |   data/bundle/je-e-21.03.02.xls
> |       output: resources/country_shapes.geojson,
> |   resources/offshore_shapes.geojson, resources/europe_shape.geojson,
> |   resources/nuts3_shapes.geojson
> |       log: logs/build_shapes.log
> |       jobid: 0
> |       resources: mem=500
> |   Traceback (most recent call last):
> |     File "/home/clara/test-egon-data/run-pypsa-eur-sec/pypsa-
> |   eur/.snakemake/scripts/tmp7bc9h0kb.build_shapes.py", line 87, in
> |   <module>
> |       import pycountry as pyc
> |   ModuleNotFoundError: No module named 'pycountry'
> |   [Tue May 11 10:25:24 2021]
> |   Error in rule build_shapes:
> |       jobid: 0
> |       output: resources/country_shapes.geojson,
> |   resources/offshore_shapes.geojson, resources/europe_shape.geojson,
> |   resources/nuts3_shapes.geojson
> |       log: logs/build_shapes.log (check log file(s) for error message)
> |   RuleException:
> |   CalledProcessError in line 122 of /home/clara/test-egon-data/run-
> |   pypsa-eur-sec/pypsa-eur/Snakefile:
> |   Command 'set -euo pipefail;  /home/clara/venv/egon-
> |   data/bin/python3.7 /home/clara/test-egon-data/run-pypsa-eur-
> |   sec/pypsa-eur/.snakemake/scripts/tmp7bc9h0kb.build_shapes.py'
> |   returned non-zero exit status 1.
> |     File "/home/clara/venv/egon-data/lib/python3.7/site-
> |   packages/snakemake/executors/__init__.py", line 2352, in run_wrapper
> |     File "/home/clara/test-egon-data/run-pypsa-eur-sec/pypsa-
> |   eur/Snakefile", line 122, in __rule_build_shapes
> |     File "/home/clara/venv/egon-data/lib/python3.7/site-
> |   packages/snakemake/executors/__init__.py", line 569, in _callback
> |     File
> |   "/home/clara/miniconda3/lib/python3.7/concurrent/futures/thread.py",
> |   line 57, in run
> |     File "/home/clara/venv/egon-data/lib/python3.7/site-
> |   packages/snakemake/executors/__init__.py", line 555, in
> |   cached_or_run
> |     File "/home/clara/venv/egon-data/lib/python3.7/site-
> |   packages/snakemake/executors/__init__.py", line 2364, in run_wrapper
> |   Shutting down, this might take some time.
> |   Exiting because a job execution failed. Look above for error message
> |   Complete log: /home/clara/test-egon-data/run-pypsa-eur-sec/pypsa-
> |   eur/.snakemake/log/2021-05-11T094608.135059.snakemake.log
> |   [Tue May 11 10:25:24 2021]
> |   Error in rule Main:
> |       jobid: 0
> |       conda-env: /home/clara/test-egon-data/run-pypsa-eur-
> |   sec/.snakemake/conda/4ea08302ffcebe8c736a25c6f2510d98
> |       shell:
> |           snakemake --directory=pypsa-eur-sec/ --snakefile=pypsa-eur-
> |   sec/Snakefile -j1 prepare_sector_networks
> |           (one of the commands exited with non-zero exit code; note
> |   that snakemake uses bash strict mode!)
> |   Shutting down, this might take some time.
> |   Exiting because a job execution failed. Look above for error message
> |   Complete log: /home/clara/test-egon-data/run-pypsa-eur-
> |   sec/.snakemake/log/2021-05-11T092558.369280.snakemake.log

**Comment by ClaraBuettner** [comment]
I still run into the  `ModuleNotFoundError: No module named 'pycountry' `- Error. 
Is it possible to activate the created conda-env from a terminal to check if the package is installed correctly?

**Comment by gnn** [comment]
> I still run into the `ModuleNotFoundError: No module named 'pycountry' `- Error.
> Is it possible to activate the created conda-env from a terminal to check if the package is installed correctly?

On my end, I could activate the environment with `conda activate pypsa-eur`. The `pycountry` package is installed inside the environment:

```
(pypsa-eur) $ conda list 2>&1 | grep pycountry
pycountry                 20.7.3             pyh9f0ad1d_0    conda-forge
```

**Comment by ClaraBuettner** [comment]
I run the branch again with the newest commits but still run into the same error. And I can not activate the conda environemnt pypsa-eur:
> 
> (base) clara@clara-LIFEBOOK-U749:~$ conda env list
>  conda environments:
> 
>/home/clara/egon-data-new-test/run-pypsa-eur-sec/.snakemake/conda/91fd94f80021084341aaf66060b6270a
>  /home/clara/egon-data-new-test/run-pypsa-eur-sec/.snakemake/conda/b39465d7463a6048afb724f038e0a4e4
> base                  *  /home/clara/miniconda3
> ...
> (base) clara@clara-LIFEBOOK-U749:~$ conda activate pypsa-eur
> Could not find conda environment: pypsa-eur

Looks the the environment doesn't get the correct name. I already had a pypsa-eur environment which I deleted before running this branch. I will check if renaming the environment helps.

**Comment by ClaraBuettner** [comment]
Another name doesn't help. I can activate the  /home/clara/egon-data-new-test/run-pypsa-eur-sec/.snakemake/conda/.... environments and there are all packages installed (also pycountry). So I assume the environment couldn't be activated correctly due to the missing name.

**Comment by ClaraBuettner** [comment]
I was able to fix my problems, pypsa-eur-sec is running but the solving fails due to memory errors. 
The missing import of packages was solved by removing the run_pypsa_eur_sec-folder and all sub folders. I think I still had some old versions of pypsa-eur and pypsa-eur-sec which were not updated automatically. In addition I changed the conda settings to not activate the `base` environment automatically. 
The current setting is still failing because 16GB RAM are not enough to solve the optimization. It works when I consider only every 10th hours, I just tested this to be sure it is only the RAM problem.

**Comment by ClaraBuettner** [comment]
When a new conda environment for pypsa-eur has to be created, some packages are installed in versions which are not suitable for our pypsa-eur-sec fork. 
I downgraded `pandas<1.3`, `xarray<0.19` and `pypsa==0.17.1` manually to solve the problems. These restrictions are also changed in the `requirements.txt` but it seems like the package versions are accessed from `pypsa-eur/envs/environemnt.yml` which is not part of egon-data or the pypsa-eur-sec fork. Unfortunately I wasn't able to find an automatic solution for this problem yet. So when anyone wants to run this branch, please make sure to use the package versions written above.

File: src/egon/data/datasets.yml [hunk 1]
```diff
@@ -306,6 +309,13 @@ peta5_0_1_ser_heat_demands:
     file_table_map:
     table: # 2015 heat demand data not stored in database
 
+hotmaps_current_policy_scenario_heat_demands_buildings:
+  original_data:
```
Comment (IlkaCu) [inline src/egon/data/datasets.yml:313]: Could be adapted to the new structure of "sources" and "targets"

File: src/egon/data/datasets/electrical_neighbours.py [hunk 2]
```diff
@@ -0,0 +1,1025 @@
...
+import zipfile
+from sqlalchemy.orm import sessionmaker
+import egon.data.datasets.etrago_setup as etrago
+
+
+class ElectricalNeighbours(Dataset):
+    def __init__(self, dependencies):
+        super().__init__(
+            name="ElectricalNeighbours",
+            version="0.0.0.dev",
```
Comment (IlkaCu) [inline src/egon/data/datasets/electrical_neighbours.py:18]: Please remove the .dev suffix

File: src/egon/data/datasets/electrical_neighbours.py [hunk 3]
```diff
@@ -0,0 +1,1025 @@
...
+        if_exists="append",
+        con=db.engine(),
+        index=False,
+    )
+
+    return new_lines
+
+
+def choose_transformer(s_nom):
+    """Select representative transformer from existing data model
```
Comment (IlkaCu) [inline src/egon/data/datasets/electrical_neighbours.py:395]: Would be helpful to extent the documentation a bit

File: src/egon/data/datasets/electrical_neighbours.py [hunk 4]
```diff
@@ -0,0 +1,1025 @@
...
+    )
+
+    # insert data
+    session = sessionmaker(bind=db.engine())()
+    for i, row in store.iterrows():
+        entry = etrago.EgonPfHvStorage(
+            scn_name="eGon2035",
+            storage_id=int(db.next_etrago_id("storage")),
+            bus=row.bus,
+            max_hours=(6 if row.carrier == "battery" else 168),
```
Comment (IlkaCu) [inline src/egon/data/datasets/electrical_neighbours.py:853]: max_hours should be taken from scenario_parameters table.

File: src/egon/data/datasets/pypsaeursec/__init__.py [hunk 5]
```diff
@@ -0,0 +1,897 @@
...
+    tasks = (run_pypsa_eur_sec, neighbor_reduction)
+else:
+    tasks = neighbor_reduction
+
+
+class PypsaEurSec(Dataset):
+    def __init__(self, dependencies):
+        super().__init__(
+            name="PypsaEurSec",
+            version="0.0.1.dev",
```
Comment (IlkaCu) [inline src/egon/data/datasets/pypsaeursec/__init__.py:894]: Remove .dev suffix before merging

File: src/egon/data/datasets/pypsaeursec/nodal_capacities.csv [hunk 6]
```diff
@@ -0,0 +1,1966 @@
+cluster,,,37
```
Comment (IlkaCu) [inline src/egon/data/datasets/pypsaeursec/nodal_capacities.csv:1]: File should be removed, in case it is not needed anymore

**Review by IlkaCu — CHANGES_REQUESTED** [review]

File: src/egon/data/datasets/pypsaeursec/__init__.py [hunk 7]
```diff
@@ -0,0 +1,897 @@
...
+        / "results"
+        / data_config["run"]
+        / "csvs/nodal_capacities.csv"
+    )
+
+    df = pd.read_csv(target_file, skiprows=5)
+    df.columns = ["component", "country", "carrier", "p_nom"]
+    df["scn_name"] = "eGon100RE"
+
+    # Todo: connection to bus!
```
[the remark this answers is not in the document]
Comment (ClaraBuettner) [reply src/egon/data/datasets/pypsaeursec/__init__.py]: This function was not used at all. I updated it and removed the todos.

File: src/egon/data/datasets/electrical_neighbours.py (hunk 4 above)
Comment (ClaraBuettner) [reply src/egon/data/datasets/electrical_neighbours.py]: For now I just updated the max_hours for pumped hydro storages. As soon as #549 is merged, I will update this.

**Review by IlkaCu — APPROVED** [review]

**Pull request merged**